### ISSUE TITLE
ksh remains in "noecho" mode forever

### ISSUE THREAD

**Issue by vmihalko** [ksh93/ksh#672]
### Detailed description of the problem:
https://bugzilla.redhat.com/show_bug.cgi?id=2185030
### KSH version:
ksh 93u+m/1.0.6 (latest)
### Reproducer
```ksh
$ ksh
$ sudo -u root echo hello|head -1
hello
    $
```
### What happens
After running the above command, the prompt, which is also not correctly printed (it should be at the beginning of the line), typing keys show nothing because the terminal remains in "noecho" mode forever.

**Comment by McDutchie** [comment]
I am unable to reproduce this on CentOS 7.9.2009 x86_64. (I have no access to RHEL for testing.)

Are you sure this bug exists in 93u+m/1.0.6? The Red Hat bug report you linked to refers to 93u+ 2012-08-01 (ksh-20120801-257.el8) and has no indication that 93u+m was tested.

**Comment by tvalenta** [comment]
I can't reproduce it on macOS M1 either.

**Comment by vmihalko** [comment]
Sorry for providing an incomplete reproducer. I realized that I forgot to include one specific step:
1. Add "Defaults use_pty" in /etc/sudoers
2. Execute the above command (from the first comment: https://github.com/ksh93/ksh/issues/672#issue-1854845462)

I can reproduce it with ksh 93u+m/1.0.6 (latest) on fedora-38.

**Comment by McDutchie** [comment]
Reproduced on Debian 11 on amd64.

**Comment by JohnoKing** [comment]
I've tried reproducing this, and so far have found it very difficult to get the terminal stuck in noecho mode (it seems random whether it does or not). The PS1 prompt consistently prints wrongly, however. For that, I've found that it has to do with ksh's use of `socketpair`; using `set -o posix` (to force use of real pipes) or disabling the `socketpair` feature test can workaround the prompt printing bug.

**Comment by McDutchie** [comment]
Consistently working reproducer:

```sh
$ stty -echo | head -1
```

Although this disables echo mode on bash, too.

**Comment by McDutchie** [comment]
Patch to test:

```diff
diff --git a/src/cmd/ksh93/sh/jobs.c b/src/cmd/ksh93/sh/jobs.c
index eeddb6ebd..8f0a8304c 100644
--- a/src/cmd/ksh93/sh/jobs.c
+++ b/src/cmd/ksh93/sh/jobs.c
@@ -440,6 +440,8 @@ int job_reap(int sig)
 			{
 				/* Reset terminal process group to ksh's own. */
 				tcsetpgrp(JOBTTY,job.mypid);
+				/* Restore terminal state (without discarding typeahead). */
+				tty_set(job.fd,TCSADRAIN,&my_stty);
 				/*
 				 * Since SIGWINCH is sent to the current process group, ksh will not have
 				 * received it if the window size changed while waiting for the job.
```

**Comment by avih** [comment]
> Consistently working reproducer:
> 
> $ stty -echo | head -1
> 
> Although this disables echo mode on bash, too.

It's not obvious to me why this is considered a bug. This knowingly changes the terminal mode. Why should ksh override it?

Is it that ksh is expected to reset/restore the terminal state when printing the prompt, but doesn't in this test case? (and only in interactive mode?)

What if the user does other changes which they want to keep, like changing terminal tab stops?

In other words, what is the desired/expected behavior in cases where some command changes some terminal/pty things?

EDIT: and I imagine that in non-interactive mode ksh shouldn't care about the terminal state?

**Comment by McDutchie** [comment]
If the reliable stty reproducer is not considered a bug, then it would seem that @vmihalko's original reproducer should be considered an intermittent problem with sudo itself and not a problem with ksh. Apparently, sudo intermittently fails to restore terminal state when connected to a socketpair.

I think the patch above is a reasonable compromise, though. It restores the saved terminal state when an external command finished. However, as of commit 4350174a5d4acabf78f97b28c6d0ae68ec703e78, the saved terminal state is no longer invalidated after running an external command while the interactive shell bits of job control are deactivated. This means that something like `stty -echo | cat` will override the effect of `stty -echo` (as well as the `sudo` bug) by restoring saved terminal state, whereas a simple `stty -echo` remains effective because `job_wait` updates the saved terminal state with the `tty_set(-1, 0, NULL);` call on line 1461.

**Comment by McDutchie** [comment]
I gave the original bug description at https://bugzilla.redhat.com/show_bug.cgi?id=2185030 a proper re-read, and it looks like I completely misunderstood the interaction of ksh with sudo. Forget everything I said before.

The issue is that ksh, by default, only waits for the last element of a pipeline. All elements of a pipeline except the last run asynchronously. This is by design (see `case TFIL` in sh_exec() xec.c), POSIX allows it, and I'm afraid it's not going to change.

So in `sudo -u root echo hello|head -1`, ksh only waits for the `head -1` part.

sudo then fails to restore the echo mode because, since it's not waited for, it intermittently loses a race where ksh considers `head -1` finished and regains control of the terminal before sudo can restore the state.

If all this is correct, then `set -o pipefail` should be an effective workaround, because that option requires that ksh wait for all the processes in the pipeline (so it can obtain the exit status of each).

**Comment by avih** [comment]
Good catch and analysis. I tend to agree the issue is a race ("tend" because I don't claim to fully understand what sudo and ksh do here, but it does sound right).

> All elements of a pipeline except the last run asynchronously. This is by design ..., POSIX allows it

Yeah.

> I'm afraid it's not going to change.

I agree, as it's valid and historic behavior, and can have value.

However, considering some other popular interactive shells do wait for the whole pipeline to complete (tested bash, zsh, busybox-ash - all of them wait), and this can have value too, as in this case, maybe consider some option to support such behavior too, but without pipefail?

It doesn't belong in 1.0 IMO, but perhaps it could be a potential TODO item for some later stage?
